src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -14,40 +14,32 @@
 from pants.engine.isolated_process import (ExecuteProcessRequest, ExecuteProcessResult,
                                            FallibleExecuteProcessResult)
 from pants.engine.legacy.graph import TransitiveHydratedTarget
-from pants.engine.rules import optionable_rule, rule
+from pants.engine.legacy.structs import PythonTestsAdaptor
+from pants.engine.rules import UnionRule, optionable_rule, rule
 from pants.engine.selectors import Get
-from pants.rules.core.core_test_model import Status, TestResult
+from pants.rules.core.core_test_model import Status, TestResult, TestTarget
 
 
-# This class currently exists so that other rules could be added which turned a HydratedTarget into
-# a language-specific test result, and could be installed alongside run_python_test.
-# Hopefully https://github.com/pantsbuild/pants/issues/4535 should help resolve this.
-class PyTestResult(TestResult):
-  pass
-
-
-# TODO: Support deps
-# TODO: Support resources
-@rule(PyTestResult, [TransitiveHydratedTarget, PyTest])
-def run_python_test(transitive_hydrated_target, pytest):
-  target_root = transitive_hydrated_target.root
+@rule(TestResult, [PythonTestsAdaptor, TransitiveHydratedTarget, PyTest])
```

Comment:
I agree. 

I'm still very new to this all so my suggestions should be taken with a grain of salt, but it makes sense to me for a `HydratedTarget` to be the wrapper. I'm currently conceptualizing a `HydratedTarget` to have all the build graph logic it needs like dependency relations, in addition to its `TargetAdaptor`. The `Target` part of `HydratedTarget` is a key clue to me.

Reply:
`HydratedTarget` is an untyped wrapper, but the thing inside of the wrapper is actually typed, and can thus have useful fields specific to its type. But moreover, the fields of HydratedTarget are redundant... all `TargetAdaptor` subclasses have both an address and dependencies.

The use of `PythonTestTarget` is the important part here, as it is what makes this rule unambiguously the right one to run python tests with. The strategy with `@union` is that multiple members of the union will be installed, each with corresponding rules to compute their test result.

`TransitiveHydratedTarget` is awkward, I agree. Figuring out how to do (typed?) "transitive deps" of a target is an open question. It might be possible to draw inspiration from bazel in this regard, but it's unclear.

Reply:
> TransitiveHydratedTarget is awkward, I agree. Figuring out how to do (typed?) "transitive deps" of a target is an open question. It might be possible to draw inspiration from bazel in this regard, but it's unclear.

It's _also_ the case that we need to be 100% clear how codegen will work before we invest too much more time in the target API. And that might mean converting `TransitiveHydratedTarget` into something much lighter that someone would then request typed-things for (ie, change it to `TransitiveBuildFileAddresses` or something).

Reply:
Maybe an answer here is for `PythonTarget` to have either a `HydratedTarget` or a `TransitiveHydratedTarget` as a field, or to be able to get one with a `yield Get` based on it or some field of it?

Reply:
I was going to say the above^! One example/extension? of that might be to just ask for a python test-specific datatype `PythonTestTargetRequest` here, which would contain exactly the requisite fields from `TargetAdaptor` and `HydratedTarget`.

Reply:
Also,
> we need to be 100% clear how codegen will work before we invest too much more time in the target API

I don't know if I agree with this approach, in general I often find the opposite direction works extremely effectively as long as relative priorities continue to be updated and discussed. If we see benefit from some change, I personally don't think we should table that until after we are 100% sure how codegen will work. I do agree that we should absolutely avoid trying to make interfaces absolutely perfect before we have an implementation that makes them useful. Not sure how to interpret this comment.

Reply:
> Maybe an answer here is for `PythonTarget` to have either a `HydratedTarget` or a `TransitiveHydratedTarget` as a field, or to be able to get one with a `yield Get` based on it or some field of it?

This is roughly what ended up happening in #7720.